### PR TITLE
[wip] Refactor preamble toc.

### PR DESCRIPTION
**Pull request by jmcarp**
Will update tests if approach is sound. cc @cmc333333

### CODE REVIEW

**Comment by cmc333333** [comment]
Is `toc-preamble.html` missing?

**Comment by jmcarp** [comment]
Added it.

File: regulations/templates/regulations/toc-preamble.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<ol>
+  {% for item in preamble_toc %}
+  <li>
+    <a
+        href="{{item.url}}"
+        data-section-id="{{item.full_id}}">
+      {{item.title}}
+    </a>
+    {% with preamble_toc=item.children %}
+      {% include "regulations/toc-preamble.html" %}
```
Comment (cmc333333) [inline regulations/templates/regulations/toc-preamble.html]: Heads up: when we were working with Django templates before, we found that they didn't handle recursion properly; we needed to convert template names into a variable. I don't know if this is still the case, but it'd be apparent at level three (the template is compiled to include itself, so levels 1 and 2 display properly).

**Comment by cmc333333** [comment]
Direction makes sense to me

File: regulations/static/regulations/js/source/helpers.js [hunk 2]
```diff
@@ -307,10 +307,10 @@ module.exports = {
       if (type === 'preamble') {
         // Note: Document ID appears twice in preamble IDs
         // TODO: Standardize IDs and drop the `slice`
-        path = path.concat(parts.slice(1));
+        path = path.concat(parts.slice(1, 2));
       } else if (type === 'cfr') {
         path = path.concat(['cfr_changes', parts.join('-')]);
       }
-      return path;
+      return {path: path, hash: parts.join('-')};
```
Comment (cmc333333) [inline regulations/static/regulations/js/source/helpers.js]: Just noting that the docstring should change

**Comment by jmcarp** [comment]
Cleaned up, tests passing. This could use more work--reasoning about child views and their URLs is still much more complicated than it should be--but at least it (mostly) works.

File: regulations/static/regulations/js/source/helpers.js [hunk 3]
```diff
@@ -290,14 +290,23 @@ module.exports = {
...
      * @example
-     * // Returns ['preamble', '2016_02749', 'I', 'A']
      * parsePreambleId('2016_02749-preamble-2016_02749-I-A')
+     * {
+     *   docId: '2016_02749',
+     *   path: ['preamble', '2016_02749', 'I'],
+     *   hash: '2016_02749-I-A'
+     * }
      * @example
      * // Returns ['preamble', '2016_02749', 'cfr_changes', '478-32-a']
```
Comment (cmc333333) [inline regulations/static/regulations/js/source/helpers.js]: This guy should be removed, yes?

File: regulations/static/regulations/js/source/helpers.js [hunk 4]
```diff
@@ -290,14 +290,23 @@ module.exports = {
...
      * parsePreambleId('2016_02749-cfr-478-32-a-1')
+     * {
+     *    docId: '2016_02749',
+     *    path: ['preamble', '2016_02749', 'cfr_changes', '478-32-a-1'],
```
Comment (cmc333333) [inline regulations/static/regulations/js/source/helpers.js]: Not for this PR, but this'll probably need to be '478-32' for similar reasons.

**Comment by cmc333333** [comment]
Going ahead with the merge so we can start working with this. Great stuff, @jmcarp

**Pull request merged**